File: project_euler/Problem049/Problem049.py
import math
from typing import List

class Problem049:
    def solution(self) -> None:
        target, other, terms, digits = 3330, 1487, [0 for _ in range(2)], [False for _ in range(10)]
        for i in range(10000 - (target << 1) - 1, 1000, -2):
            if self.__is_prime(i) and i != other:
                terms[0] = i + target
                terms[1] = terms[0] + target
                if self.__is_prime(terms[0]) and self.__is_prime(terms[1]) and self.__check_digit(i, terms, digits):
                    print('%d%d%d' % (i, terms[0], terms[1]))
                    return

    def __is_prime(self, value: int) -> bool:
        if not (value & 1):
            return 2 == value

        border = int(math.sqrt(value))
        for i in range(3, border + 1, 2):
            if not value % i:
                return False
        return 1 != value

    def __check_digit(self, value: int, terms: List[int], digits: List[bool]) -> bool:
        digits[:] = [False for _ in range(10)]
        while value:
            digits[value % 10] = True
            value //= 10

        for term in terms:
            while term:
                if not digits[term % 10]:
                    return False
                term //= 10
        return True

File: project_euler/Problem049/Problem049.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem049() {
    target := 3330
    other := 1487
    terms := make([]int, 2)
    digits := make([]bool, 10)
    for i := 10000 - (target << 1) - 1; i > 1000; i -= 2 {
        if isPrime(i) && i != other {
            terms[0] = i + target
            terms[1] = terms[0] + target
            if isPrime(terms[0]) && isPrime(terms[1]) && checkDigit(i, terms, digits) {
                fmt.Printf("%d%d%d\n", i, terms[0], terms[1])
                return
            }
        }
    }
}

func isPrime(value int) bool {
    if 0 == (value & 1) {
        return 2 == value
    }

    border := int(math.Sqrt(float64(value)))
    for i := 3; i <= border; i += 2 {
        if 0 == value % i {
            return false
        }
    }
    return 1 != value
}

func checkDigit(value int, terms []int, digits []bool) bool {
    for i := 0; i < 10; i++ {
        digits[i] = false
    }
    for 0 != value {
        digits[value % 10] = true
        value /= 10
    }

    for _, term := range terms {
        for 0 != term {
            if !digits[term % 10] {
                return false
            }
            term /= 10
        }
    }
    return true
}
